### PR TITLE
📑 Add `literalinclude` directive

### PR DESCRIPTION
**Pull request by rowanc1**
This supports a new directive, `literalinclude` as well as all of the additional options across RST and Sphinx.

RST documentation:
 - https://docutils.sourceforge.io/docs/ref/rst/directives.html#including-an-external-document-fragment

Sphinx documentation:
 - https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude

See executablebooks/jupyterlab-myst#189 and https://github.com/orgs/executablebooks/discussions/1026

There are a few more tweaks that will be necessary to bring this to `jupyterlab-myst` to expose the transform and extend to not having to read from disk, etc.

cc @nthiery @matthewfeickert @fperez who were asking for this over the last few months.

### CODE REVIEW

File: packages/myst-directives/src/code.ts [hunk 1]
```diff
@@ -1,56 +1,83 @@
...
+  if (options?.['lineno-start'] != null && options?.['number-lines'] != null) {
+    fileWarn(vfile, 'Cannot use both "lineno-start" and "number-lines"', {
+      source: 'code-cell:tags',
+    });
+  }
+  const emphasizeLines = parseEmphasizeLines(options?.['emphasize-lines'] as string | undefined);
+  const numberLines = options?.['number-lines'] as number | undefined;
+  // Only include this in mdast if it is `true`
+  const showLineNumbers =
+    options?.linenos || options?.['lineno-start'] || options?.['lineno-match'] || numberLines
```
Comment (rowanc1) [inline packages/myst-directives/src/code.ts:31]: This fixes a portion of the bug in:
- https://github.com/executablebooks/jupyterlab-myst/issues/174

Line number is now turned on if `lineno-start` is included.

**Review by fwkoch — APPROVED** [review]
This looks great - I didn't dig through all the logic around line numbers, etc, in detail, but testing and documentation both look thorough 👍 

Despite all the logic complexity added, there was nice simplification around previously duplicated options/directives. My only comments are non-critical.

File: packages/myst-directives/src/code.ts [hunk 2]
```diff
@@ -1,56 +1,83 @@
...
+}
+
+/** This function parses both sphinx and RST code-block options */
+export function getCodeBlockOptions(
+  options: DirectiveData['options'],
+  vfile: VFile,
+): Pick<Code, 'emphasizeLines' | 'showLineNumbers' | 'startingLineNumber'> {
+  if (options?.['lineno-start'] != null && options?.['number-lines'] != null) {
+    fileWarn(vfile, 'Cannot use both "lineno-start" and "number-lines"', {
+      source: 'code-cell:tags',
```
[the remark this answers is not in the document]
Comment (rowanc1) [reply packages/myst-directives/src/code.ts:24]: Ah. copy-paste error!

File: packages/myst-cli/src/transforms/include.ts [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
-    const content = fs.readFileSync(file).toString();
-    const children = parseMyst(session, content, filename).children as GenericNode[];
-    node.children = children;
-  });
+    return fs.readFileSync(fullFile).toString();
+  };
+  const parseContent = (filename: string, content: string) => {
+    return parseMyst(session, content, filename).children;
+  };
+  includeDirectiveTransform(tree, vfile, { loadFile, parseContent });
```
[the remark this answers is not in the document]
Comment (rowanc1) [reply packages/myst-cli/src/transforms/include.ts:28]: Yeah, which will change when we do it not using the file system!

File: packages/myst-directives/src/code.ts [hunk 4]
```diff
@@ -1,56 +1,83 @@
...
+    startingLineNumber = undefined;
+  }
+  return {
+    emphasizeLines,
+    showLineNumbers,
+    startingLineNumber,
+  };
+}
+
+export const CODE_DIRECTIVE_OPTIONS: DirectiveSpec['options'] = {
```
[the remark this answers is not in the document]
Comment (rowanc1) [reply packages/myst-directives/src/code.ts:48]: My thinking was to move that to it's own reusable option in the future (but not now...!)

**Comment by nthiery** [comment]
I confirm that the dev version of myst supports smoothly the literalincludes
in my documents.

Since I could not yet test it with JupyterLab, I actually started `mystmd` for the
first time on a real project of mine. Wow, this is sooo fast. True real time
rendering. This a game changer for authoring.

Guys, it's amazing what you are building!!!

Can't wait to switch over from Jupyter-book for my whole toolchain.

**Comment by rowanc1** [comment]
⚡️💨🚀

Really glad you are up and running with the CLI -- we will get it over to Jupyter soon!

**Comment by fperez** [comment]
Pinging also @minrk who may have content using these.

Haven't had a chance to test it yet, but wanted to thank @rowanc1 & team for listening to our input! This is looking wonderful, will provide concrete feedback as soon as I have any. Hugely apreciated!

**Comment by nthiery** [comment]
Dear @rowanc1,

There seems to be a little glitch in the tex export when literal-including indented code.
Here is an example:

I have a file `mystere-correction.cpp` containing:
```
void mystereTest() {
    /// BEGIN mystereTests
    CHECK( mystere(0) == ""             );
    CHECK( mystere(1) == "*\n"          );
    CHECK( mystere(2) == "*\n**\n"      );
    CHECK( mystere(3) == "*\n**\n***\n" );
    /// END mystereTests
}
```
It's included as such:
```
:::{literalinclude} mystere-correction.cpp
:start-after: BEGIN mystereTests
:end-before: END mystereTests
:::
```

The generated TeX looks as such; mind the indentation of the first line:
```
\begin{verbatim}
CHECK( mystere(0) == ""             );
    CHECK( mystere(1) == "*\n"          );
    CHECK( mystere(2) == "*\n**\n"      );
    CHECK( mystere(3) == "*\n**\n***\n" );
\end{verbatim}
```

Let me know if more detailed instructions to reproduce would be helpful.

The generated html is fine.

**Comment by fwkoch** [comment]
@nthiery - I pulled your above comment to a separate issue so it wasn't lost: https://github.com/executablebooks/mystmd/issues/614

**Pull request merged**